src/ControlFlow/ControlFlowGraph.php

```diff
@@ -0,0 +1,344 @@
+            if ($statement->stmts) {
+                $this->passNodes($statement->stmts, $this->root);
+            }
+        }
+    }
+
+    /**
+     * @param \PhpParser\Node[] $nodes
+     * @param Block $block
+     */
+    protected function passNodes(array $nodes, Block $block)
+    {
+        foreach ($nodes as $stmt) {
+            switch (get_class($stmt)) {
+                case \PhpParser\Node\Stmt\Goto_::class:
+                    if (isset($this->labels[$stmt->name])) {
+                        $block->addChildren(
+                            new Node\JumpNode($this->labels[$stmt->name])
+                        );
+                    } else {
+                        $this->unresolvedGotos[] = $stmt;
+                    }
+                    break;
+                case \PhpParser\Node\Expr\Assign::class:
+                    $this->passAssign($stmt, $block);
+                    break;
+                case \PhpParser\Node\Stmt\Return_::class:
+                    $this->passReturn($stmt, $block);
+                    break;
+                case \PhpParser\Node\Stmt\For_::class:
+                    $block = $this->passFor($stmt, $block);
+                    break;
+                case \PhpParser\Node\Stmt\If_::class:
+                    $block = $this->passIf($stmt, $block);
+                    break;
+                case \PhpParser\Node\Stmt\While_::class:
+                    $block = $this->passWhile($stmt, $block);
+                    break;
+                case \PhpParser\Node\Stmt\Do_::class:
+                    $block = $this->passDo($stmt, $block);
+                    break;
+                case \PhpParser\Node\Stmt\Throw_::class:
+                    $this->passThrow($stmt, $block);
+                    break;
+                case \PhpParser\Node\Expr\Exit_::class:
+                    $block->addChildren(new Node\ExitNode());
+                    break;
+                case \PhpParser\Node\Stmt\Label::class:
+                    $block = $this->createNewBlockIfNeeded($block);
+                    $block->label = $stmt->name;
+                    $this->labels[$block->label] = $block;
+                    break;
+                case \PhpParser\Node\Stmt\TryCatch::class:
+                    $block = $this->passTryCatch($stmt, $block);
+                    break;
+                case \PhpParser\Node\Stmt\Nop::class:
+                    // ignore commented code
+                    break;
+                default:
+                    echo 'Unimplemented ' . get_class($stmt) . PHP_EOL;
+                    break;
+            }
+        }
+    }
+
+    /**
+     * If current block is not empty, lets create a new one
+     *
+     * @param Block $block
+     * @return Block
+     */
+    protected function createNewBlockIfNeeded(Block $block)
+    {
+        if ($block->getChildren()) {
+            $block->setExit(
+                $block = new Block($this->lastBlockId++)
```

Comment:
[Assignment in sub-expressions](https://sonarqube.com/coding_rules#rule_key=c%3AAssignmentInSubExpression)

This seems not to be even circular reference, is it intentional ?

Maybe the same as

```php
    protected function createNewBlockIfNeeded(Block $block)
    {
        if ($block->getChildrens()) {
            $block = new Block($this->lastBlockId++);
        }

        return $block;
    }
```